In [1]:
import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings('ignore')

#### importing the excel sheets:

In [2]:
data_service = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Service Data')
data_Employee = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Employee')
data_service_type = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Service Type')
data_channel_type= pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Channel Type')
data_fault_type = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Fault Type')
data_location = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Location')
data_resources = pd.read_excel('Zentel Network Service Ticket Data.xlsx',sheet_name='Resources')

In [3]:
data_service.head()

Report ID Report Channel Customer Name State Key  \
0  AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1  AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2  AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3    AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4    AXA-20201231-1105-HSE           CH01          Arik    NGS003   

     Ticket Open Time    Ticket Resp Time      Issue Res Time     Fault Type  \
0 2020-12-31 17:07:04 2020-12-31 17:10:20 2020-12-31 20:44:42            NaN   
1 2020-12-30 19:04:06 2020-12-30 19:13:18 2020-12-30 19:46:30  Customers End   
2 2020-12-29 18:28:04 2020-12-29 18:37:52 2020-12-29 20:10:11  Customers End   
3 2020-12-29 14:49:16 2020-12-29 14:58:23 2020-12-29 16:03:56    Line damage   
4 2020-12-31 21:43:03 2020-12-31 21:47:39 2020-12-31 22:20:58  Customers End   

   Operator    Ticket Close Time Ticket Status Business Status  
0     Tunde                             Active            Open  
1    Sekina  2021-01-01 13:27:17     Completed          Closed  
2  Sherifat  2020-12-30 13:20:43     Completed          Closed  
3  Kingsley  2020-12-29 21:34:31     Completed          Closed  
4      Kola  2021-01-02 06:54:01     Completed          Closed

In [4]:
data_service.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5998 entries, 0 to 5997
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Report ID          5998 non-null   object        
 1   Report Channel     5998 non-null   object        
 2   Customer Name      5998 non-null   object        
 3   State Key          5998 non-null   object        
 4   Ticket Open Time   5998 non-null   datetime64[ns]
 5   Ticket Resp Time   5998 non-null   datetime64[ns]
 6   Issue Res Time     5998 non-null   datetime64[ns]
 7   Fault Type         5472 non-null   object        
 8   Operator           5998 non-null   object        
 9   Ticket Close Time  5998 non-null   object        
 10  Ticket Status      5998 non-null   object        
 11  Business Status    5998 non-null   object        
dtypes: datetime64[ns](3), object(9)
memory usage: 562.4+ KB


In [5]:
data_service.isnull().sum()

Report ID              0
Report Channel         0
Customer Name          0
State Key              0
Ticket Open Time       0
Ticket Resp Time       0
Issue Res Time         0
Fault Type           526
Operator               0
Ticket Close Time      0
Ticket Status          0
Business Status        0
dtype: int64

##### converting date related columns to datetime:

In [6]:
data_service['Ticket Open Time'] = pd.to_datetime(data_service['Ticket Open Time'])
data_service['Ticket Resp Time'] = pd.to_datetime(data_service['Ticket Resp Time'])
data_service['Issue Res Time'] = pd.to_datetime(data_service['Issue Res Time'])
#data_service['Ticket Close Time'] = pd.to_datetime(data_service['Ticket Close Time'])

Ticket Close Time column contained string data;

Investigating into it:

In [7]:
data_service['Ticket Close Time'].unique()

array([' ', datetime.datetime(2021, 1, 1, 13, 27, 17),
       datetime.datetime(2020, 12, 30, 13, 20, 43), ...,
       datetime.datetime(2020, 12, 10, 16, 17, 37, 71000),
       datetime.datetime(2020, 12, 8, 13, 39, 54, 601000),
       datetime.datetime(2020, 12, 1, 19, 3, 24, 649000)], dtype=object)

In [8]:
data_service['Ticket Close Time'].value_counts()

                              639
2020-12-30 13:28:04             2
2021-01-01 07:25:36             2
2021-01-01 07:17:05             2
2021-01-01 07:03:33             2
                             ... 
2020-12-24 18:06:17.165000      1
2020-12-10 18:04:50.512000      1
2020-12-25 13:50:50.856000      1
2020-12-23 21:30:49.243000      1
2020-12-01 19:03:24.649000      1
Name: Ticket Close Time, Length: 5355, dtype: int64

639 entries (12% of entire data) was whitespace in the Ticket Close Time column;

In [9]:
# investigating into the cases
data_service[data_service['Ticket Close Time'] == ' ']

Report ID Report Channel Customer Name State Key  \
0     AXA-20201231-1101-WLESS           CH01        Access    NGS001   
14      AXA-20201231-1115-HSE           CH01         Bates    NGS004   
25    AXA-20201231-1126-WLESS           CH01       Cassava    NGS001   
60    AXA-20201229-1161-WLESS           CH01          Dang    NGS012   
61      AXA-20201229-1162-HSE           CH01          Dang    NGS012   
...                       ...            ...           ...       ...   
5889  AXA-20201219-6990-VOICE           CH05        BetSan    NGS007   
5901    AXA-20201214-7002-HSE           CH01     TwinsDeck    NGS004   
5937  AXA-20201227-7038-WiFi6           CH05       BetBoss    NGS027   
5965   AXA-20201223-7066-MCON           CH04        BetSan    NGS012   
5968  AXA-20201207-7069-WiFi9           CH03         Union    NGS001   

            Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0    2020-12-31 17:07:04.000 2020-12-31 17:10:20.000 2020-12-31 20:44:42.000   
14   2020-12-31 14:49:04.000 2020-12-31 14:58:11.000 2020-12-31 14:59:26.000   
25   2020-12-31 16:49:03.000 2020-12-31 16:50:37.000 2020-12-31 20:54:02.000   
60   2020-12-29 20:58:06.000 2020-12-29 21:06:15.000 2020-12-29 22:13:32.000   
61   2020-12-29 21:01:03.000 2020-12-29 21:01:44.000 2020-12-30 01:41:21.000   
...                      ...                     ...                     ...   
5889 2020-12-19 10:40:24.283 2020-12-19 10:40:24.283 2020-12-19 10:40:24.283   
5901 2020-12-14 14:14:35.736 2020-12-14 14:14:35.736 2020-12-14 14:14:35.736   
5937 2020-12-27 09:34:20.983 2020-12-27 09:34:20.983 2020-12-27 09:34:20.983   
5965 2020-12-23 14:07:58.144 2020-12-23 14:07:58.144 2020-12-23 14:07:58.144   
5968 2020-12-07 08:22:18.750 2020-12-07 08:22:18.750 2020-12-07 08:22:18.750   

            Fault Type  Operator Ticket Close Time Ticket Status  \
0                  NaN     Tunde                          Active   
14                 NaN      John                          Active   
25                 NaN     Atiku                          Active   
60                 NaN     Tunde                          Active   
61                 NaN    Habeeb                      Terminated   
...                ...       ...               ...           ...   
5889  hardware failure    Habeeb                          Active   
5901     Customers End  Kingsley                          Active   
5937               NaN     Atiku                          Active   
5965     Customers End  Jennifer                          Active   
5968       Line damage      Akin                          Active   

     Business Status  
0               Open  
14              Open  
25              Open  
60              Open  
61        Terminated  
...              ...  
5889            Open  
5901            Open  
5937            Open  
5965            Open  
5968            Open  

[639 rows x 12 columns]

We could fill the empty spaces with the average close time of each day the ticket was opened , but for the purpose of this analysis, we didn't have a heavy dependence on the Ticket Close Time , so we ignored it

-------------------------------------------------------

#### creating timed differenced columns for Response & Resolution;

In [10]:
data_service['Response_time_diff'] =  data_service['Ticket Resp Time'] - data_service['Ticket Open Time']
data_service['Resolution_time_diff'] = data_service['Issue Res Time'] - data_service['Ticket Resp Time']

In [11]:
data_service.head()

Report ID Report Channel Customer Name State Key  \
0  AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1  AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2  AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3    AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4    AXA-20201231-1105-HSE           CH01          Arik    NGS003   

     Ticket Open Time    Ticket Resp Time      Issue Res Time     Fault Type  \
0 2020-12-31 17:07:04 2020-12-31 17:10:20 2020-12-31 20:44:42            NaN   
1 2020-12-30 19:04:06 2020-12-30 19:13:18 2020-12-30 19:46:30  Customers End   
2 2020-12-29 18:28:04 2020-12-29 18:37:52 2020-12-29 20:10:11  Customers End   
3 2020-12-29 14:49:16 2020-12-29 14:58:23 2020-12-29 16:03:56    Line damage   
4 2020-12-31 21:43:03 2020-12-31 21:47:39 2020-12-31 22:20:58  Customers End   

   Operator    Ticket Close Time Ticket Status Business Status  \
0     Tunde                             Active            Open   
1    Sekina  2021-01-01 13:27:17     Completed          Closed   
2  Sherifat  2020-12-30 13:20:43     Completed          Closed   
3  Kingsley  2020-12-29 21:34:31     Completed          Closed   
4      Kola  2021-01-02 06:54:01     Completed          Closed   

  Response_time_diff Resolution_time_diff  
0    0 days 00:03:16      0 days 03:34:22  
1    0 days 00:09:12      0 days 00:33:12  
2    0 days 00:09:48      0 days 01:32:19  
3    0 days 00:09:07      0 days 01:05:33  
4    0 days 00:04:36      0 days 00:33:19

In [12]:
data_service.shape

(5998, 14)

Mean response time;

In [13]:
data_service['Response_time_diff'].mean()

Timedelta('0 days 00:04:33.683728076')

#### Investigating cases where there was no observed Response and Resolution time difference 

In [14]:
no_time_diff = data_service[data_service['Response_time_diff'] == data_service['Response_time_diff'].min()]
no_time_diff.sort_values(by='Ticket Open Time',inplace=True)
no_time_diff.reset_index(drop=True,inplace=True)

In [15]:
no_time_diff

Report ID Report Channel Customer Name State Key  \
0       AXA-20201201-6732-HSE           CH03           Dan    NGS032   
1       AXA-20201201-6976-HSE           CH04         ROCAD    NGS017   
2    AXA-20201201-6776-EVOICE           CH04     ExcluDive    NGS004   
3    AXA-20201201-6128-WiFi11           CH03        Allied    NGS024   
4       AXA-20201201-6173-HSE           CH03        BetSan    NGS020   
..                        ...            ...           ...       ...   
969   AXA-20201228-6781-VOICE           CH03    PrinceLine    NGS029   
970   AXA-20201228-6938-WiFi8           CH01          PWAN    NGS011   
971   AXA-20201228-6891-VOICE           CH04        Mikano    NGS011   
972    AXA-20201228-6470-C100           CH01     Honeywell    NGS008   
973   AXA-20201228-6982-VOICE           CH02     Chanrai's    NGS008   

           Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0   2020-12-01 08:09:12.134 2020-12-01 08:09:12.134 2020-12-01 08:09:12.134   
1   2020-12-01 08:30:13.603 2020-12-01 08:30:13.603 2020-12-01 08:30:13.603   
2   2020-12-01 08:47:24.374 2020-12-01 08:47:24.374 2020-12-01 08:47:24.374   
3   2020-12-01 09:08:12.602 2020-12-01 09:08:12.602 2020-12-01 09:08:12.602   
4   2020-12-01 09:11:57.896 2020-12-01 09:11:57.896 2020-12-01 09:11:57.896   
..                      ...                     ...                     ...   
969 2020-12-28 15:05:00.791 2020-12-28 15:05:00.791 2020-12-28 15:05:00.791   
970 2020-12-28 15:41:18.152 2020-12-28 15:41:18.152 2020-12-28 15:41:18.152   
971 2020-12-28 15:44:58.428 2020-12-28 15:44:58.428 2020-12-28 15:44:58.428   
972 2020-12-28 15:54:21.426 2020-12-28 15:54:21.426 2020-12-28 15:54:21.426   
973 2020-12-28 15:56:33.412 2020-12-28 15:56:33.412 2020-12-28 15:56:33.412   

                   Fault Type  Operator           Ticket Close Time  \
0                 Line damage  Jonathan  2020-12-01 12:22:57.134000   
1                         NaN    Sekina  2020-12-01 12:12:48.603000   
2               Customers End      Mary  2020-12-01 14:04:04.374000   
3               Customers End  Jonathan  2020-12-01 13:54:02.602000   
4                 Line damage     Tunde  2020-12-01 13:00:40.896000   
..                        ...       ...                         ...   
969               Line damage     Vitus  2020-12-28 18:50:25.791000   
970             Customers End  Jennifer  2020-12-28 20:17:12.152000   
971  Support network failures  Kingsley  2020-12-28 21:12:41.428000   
972               Line damage      Sola  2020-12-28 21:16:37.426000   
973             Customers End  Jennifer  2020-12-28 21:43:27.412000   

    Ticket Status Business Status Response_time_diff Resolution_time_diff  
0       Completed          Closed             0 days               0 days  
1       Completed          Closed             0 days               0 days  
2       Completed          Closed             0 days               0 days  
3       Completed          Closed             0 days               0 days  
4       Completed          Closed             0 days               0 days  
..            ...             ...                ...                  ...  
969     Completed          Closed             0 days               0 days  
970     Completed          Closed             0 days               0 days  
971     Completed          Closed             0 days               0 days  
972     Completed          Closed             0 days               0 days  
973     Completed          Closed             0 days               0 days  

[974 rows x 14 columns]

##### 974 entries showed 0 time difference between response and resolution

Investigating further ;

In [16]:
no_time_diff['Report ID'].value_counts()

AXA-20201201-6732-HSE       1
AXA-20201220-6842-CBS       1
AXA-20201219-6335-HSE       1
AXA-20201219-7012-WiFi11    1
AXA-20201219-6262-WiFi4     1
                           ..
AXA-20201210-6182-WiFi6     1
AXA-20201210-6763-EVOICE    1
AXA-20201210-6552-HSE       1
AXA-20201210-6230-WiFi5     1
AXA-20201228-6982-VOICE     1
Name: Report ID, Length: 974, dtype: int64

all report IDs are unique, no duplicated report;

In [17]:
no_time_diff['Customer Name'].value_counts()

BetSan           107
ShopLite          65
Transnational     56
Jiji              52
International     39
                ... 
Systemtech         1
Budelak            1
TITech             1
Clay PLC           1
Shell              1
Name: Customer Name, Length: 114, dtype: int64

In [18]:
no_time_diff[no_time_diff['Customer Name'] == 'BetSan']

Report ID Report Channel Customer Name State Key  \
4      AXA-20201201-6173-HSE           CH03        BetSan    NGS020   
6      AXA-20201201-6812-ICS           CH05        BetSan    NGS022   
11   AXA-20201201-6940-VOICE           CH03        BetSan    NGS012   
20   AXA-20201201-6592-WiFi4           CH04        BetSan    NGS003   
26   AXA-20201201-7098-VOICE           CH01        BetSan    NGS018   
..                       ...            ...           ...       ...   
933  AXA-20201227-7080-WiFi9           CH06        BetSan    NGS033   
944  AXA-20201227-6695-WLESS           CH04        BetSan    NGS020   
946  AXA-20201228-6299-WiFi8           CH02        BetSan    NGS019   
951   AXA-20201228-7058-MCON           CH03        BetSan    NGS005   
960  AXA-20201228-6651-WLESS           CH05        BetSan    NGS025   

           Ticket Open Time        Ticket Resp Time          Issue Res Time  \
4   2020-12-01 09:11:57.896 2020-12-01 09:11:57.896 2020-12-01 09:11:57.896   
6   2020-12-01 09:17:51.432 2020-12-01 09:17:51.432 2020-12-01 09:17:51.432   
11  2020-12-01 10:26:25.882 2020-12-01 10:26:25.882 2020-12-01 10:26:25.882   
20  2020-12-01 12:30:19.424 2020-12-01 12:30:19.424 2020-12-01 12:30:19.424   
26  2020-12-01 14:46:10.649 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649   
..                      ...                     ...                     ...   
933 2020-12-27 13:54:55.881 2020-12-27 13:54:55.881 2020-12-27 13:54:55.881   
944 2020-12-27 15:53:10.677 2020-12-27 15:53:10.677 2020-12-27 15:53:10.677   
946 2020-12-28 08:16:45.205 2020-12-28 08:16:45.205 2020-12-28 08:16:45.205   
951 2020-12-28 09:21:06.689 2020-12-28 09:21:06.689 2020-12-28 09:21:06.689   
960 2020-12-28 13:09:31.816 2020-12-28 13:09:31.816 2020-12-28 13:09:31.816   

          Fault Type  Operator           Ticket Close Time Ticket Status  \
4        Line damage     Tunde  2020-12-01 13:00:40.896000     Completed   
6      Customers End     Atiku  2020-12-01 14:03:43.432000     Completed   
11   Routine Service  Sherifat  2020-12-01 14:13:49.882000     Completed   
20       Line damage  Jennifer  2020-12-01 17:12:58.424000     Completed   
26     Customers End      Boye  2020-12-01 19:03:24.649000     Completed   
..               ...       ...                         ...           ...   
933    Customers End      John  2020-12-27 17:49:50.881000     Completed   
944  Routine Service      Mary  2020-12-27 21:43:31.677000     Completed   
946  Routine Service      John  2020-12-28 13:49:27.205000     Completed   
951    Customers End     Tunde  2020-12-28 12:42:21.689000     Completed   
960      Line damage      Bola  2020-12-28 18:33:21.816000     Completed   

    Business Status Response_time_diff Resolution_time_diff  
4            Closed             0 days               0 days  
6            Closed             0 days               0 days  
11           Closed             0 days               0 days  
20           Closed             0 days               0 days  
26           Closed             0 days               0 days  
..              ...                ...                  ...  
933          Closed             0 days               0 days  
944          Closed             0 days               0 days  
946          Closed             0 days               0 days  
951          Closed             0 days               0 days  
960          Closed             0 days               0 days  

[107 rows x 14 columns]

##### Checking to see if there's a notable difference in ticket open time and close time;

In [19]:
no_time_diff[no_time_diff['Ticket Open Time'] == no_time_diff['Ticket Close Time']]

Empty DataFrame
Columns: [Report ID, Report Channel, Customer Name, State Key, Ticket Open Time, Ticket Resp Time, Issue Res Time, Fault Type, Operator, Ticket Close Time, Ticket Status, Business Status, Response_time_diff, Resolution_time_diff]
Index: []

**conclusion** :
Clearly the observed anomaly wasn't observed in the Ticket Close Time Column.
The 974 entries can be as a result of a data quality issue and seeing as this represented 16.2% of the data, it needs to be corrected as it's effects can skew up analysis results;

------------------------------------------------------

#### Further Exploration on the provided dataset;

In [20]:
data_service['Fault Type'].value_counts()

Customers End               3151
Line damage                 1195
Routine Service              622
hardware failure             237
Support network failures     172
Traffic Spikes                31
Router Failure                31
Power Loss                    23
Intrusions                    10
Name: Fault Type, dtype: int64

Extracting hour of day In other to drill down into recommendations for **TAT** ( Turn - Around - Time) :

In [21]:
data_service['Hour_of_day'] = data_service['Ticket Open Time'].dt.hour

Feature to know the hours taken to resolve a ticket;

In [22]:
data_service['resolution_hour'] = data_service['Resolution_time_diff'].dt.total_seconds()/3600

**formula** to replicate resolution matrix:

In [23]:
def hour_conversion(row_value):
    if row_value < 0.5:
        return 'Excellent'
    elif row_value >= 0.5 and row_value <= 1:
        return 'Good'
    elif row_value > 1 and row_value <= 3:
        return 'Fair'
    elif row_value > 3:
        return 'Critical'

In [24]:
hour_conversion(4.368056)

'Critical'

In [25]:
data_service['Resolution_class'] = data_service['resolution_hour'].apply(hour_conversion)

In [26]:
data_service['Resolution_class'].value_counts()

Critical     2504
Fair         1680
Excellent    1528
Good          286
Name: Resolution_class, dtype: int64

In [27]:
data_service

Report ID Report Channel Customer Name State Key  \
0     AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1     AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2     AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3       AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4       AXA-20201231-1105-HSE           CH01          Arik    NGS003   
...                       ...            ...           ...       ...   
5993  AXA-20201204-7094-WiFi5           CH06        Lemway    NGS003   
5994  AXA-20201226-7095-WLESS           CH05        BetSan    NGS004   
5995  AXA-20201210-7096-WiFi6           CH06    ValueNaija    NGS029   
5996  AXA-20201208-7097-WLESS           CH03         Sahad    NGS007   
5997  AXA-20201201-7098-VOICE           CH01        BetSan    NGS018   

            Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0    2020-12-31 17:07:04.000 2020-12-31 17:10:20.000 2020-12-31 20:44:42.000   
1    2020-12-30 19:04:06.000 2020-12-30 19:13:18.000 2020-12-30 19:46:30.000   
2    2020-12-29 18:28:04.000 2020-12-29 18:37:52.000 2020-12-29 20:10:11.000   
3    2020-12-29 14:49:16.000 2020-12-29 14:58:23.000 2020-12-29 16:03:56.000   
4    2020-12-31 21:43:03.000 2020-12-31 21:47:39.000 2020-12-31 22:20:58.000   
...                      ...                     ...                     ...   
5993 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525   
5994 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377   
5995 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071   
5996 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601   
5997 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649   

         Fault Type  Operator           Ticket Close Time Ticket Status  \
0               NaN     Tunde                                    Active   
1     Customers End    Sekina         2021-01-01 13:27:17     Completed   
2     Customers End  Sherifat         2020-12-30 13:20:43     Completed   
3       Line damage  Kingsley         2020-12-29 21:34:31     Completed   
4     Customers End      Kola         2021-01-02 06:54:01     Completed   
...             ...       ...                         ...           ...   
5993    Line damage      Sheg  2020-12-04 12:53:29.525000     Completed   
5994  Customers End    Brenda  2020-12-26 16:58:54.377000     Completed   
5995  Customers End      Sola  2020-12-10 16:17:37.071000     Completed   
5996    Line damage  Jonathan  2020-12-08 13:39:54.601000     Completed   
5997  Customers End      Boye  2020-12-01 19:03:24.649000     Completed   

     Business Status Response_time_diff Resolution_time_diff  Hour_of_day  \
0               Open    0 days 00:03:16      0 days 03:34:22           17   
1             Closed    0 days 00:09:12      0 days 00:33:12           19   
2             Closed    0 days 00:09:48      0 days 01:32:19           18   
3             Closed    0 days 00:09:07      0 days 01:05:33           14   
4             Closed    0 days 00:04:36      0 days 00:33:19           21   
...              ...                ...                  ...          ...   
5993          Closed    0 days 00:00:00      0 days 00:00:00            9   
5994          Closed    0 days 00:00:00      0 days 00:00:00           11   
5995          Closed    0 days 00:00:00      0 days 00:00:00           12   
5996          Closed    0 days 00:00:00      0 days 00:00:00            9   
5997          Closed    0 days 00:00:00      0 days 00:00:00           14   

      resolution_hour Resolution_class  
0            3.572778         Critical  
1            0.553333             Good  
2            1.538611             Fair  
3            1.092500             Fair  
4            0.555278             Good  
...               ...              ...  
5993         0.000000        Excellent  
5994         0.000000        Excellent  

feature to showing total response seconds for a report:

In [28]:
data_service['total_response_seconds'] = data_service['Response_time_diff'].dt.total_seconds()

In [29]:
data_service

Report ID Report Channel Customer Name State Key  \
0     AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1     AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2     AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3       AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4       AXA-20201231-1105-HSE           CH01          Arik    NGS003   
...                       ...            ...           ...       ...   
5993  AXA-20201204-7094-WiFi5           CH06        Lemway    NGS003   
5994  AXA-20201226-7095-WLESS           CH05        BetSan    NGS004   
5995  AXA-20201210-7096-WiFi6           CH06    ValueNaija    NGS029   
5996  AXA-20201208-7097-WLESS           CH03         Sahad    NGS007   
5997  AXA-20201201-7098-VOICE           CH01        BetSan    NGS018   

            Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0    2020-12-31 17:07:04.000 2020-12-31 17:10:20.000 2020-12-31 20:44:42.000   
1    2020-12-30 19:04:06.000 2020-12-30 19:13:18.000 2020-12-30 19:46:30.000   
2    2020-12-29 18:28:04.000 2020-12-29 18:37:52.000 2020-12-29 20:10:11.000   
3    2020-12-29 14:49:16.000 2020-12-29 14:58:23.000 2020-12-29 16:03:56.000   
4    2020-12-31 21:43:03.000 2020-12-31 21:47:39.000 2020-12-31 22:20:58.000   
...                      ...                     ...                     ...   
5993 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525   
5994 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377   
5995 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071   
5996 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601   
5997 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649   

         Fault Type  Operator           Ticket Close Time Ticket Status  \
0               NaN     Tunde                                    Active   
1     Customers End    Sekina         2021-01-01 13:27:17     Completed   
2     Customers End  Sherifat         2020-12-30 13:20:43     Completed   
3       Line damage  Kingsley         2020-12-29 21:34:31     Completed   
4     Customers End      Kola         2021-01-02 06:54:01     Completed   
...             ...       ...                         ...           ...   
5993    Line damage      Sheg  2020-12-04 12:53:29.525000     Completed   
5994  Customers End    Brenda  2020-12-26 16:58:54.377000     Completed   
5995  Customers End      Sola  2020-12-10 16:17:37.071000     Completed   
5996    Line damage  Jonathan  2020-12-08 13:39:54.601000     Completed   
5997  Customers End      Boye  2020-12-01 19:03:24.649000     Completed   

     Business Status Response_time_diff Resolution_time_diff  Hour_of_day  \
0               Open    0 days 00:03:16      0 days 03:34:22           17   
1             Closed    0 days 00:09:12      0 days 00:33:12           19   
2             Closed    0 days 00:09:48      0 days 01:32:19           18   
3             Closed    0 days 00:09:07      0 days 01:05:33           14   
4             Closed    0 days 00:04:36      0 days 00:33:19           21   
...              ...                ...                  ...          ...   
5993          Closed    0 days 00:00:00      0 days 00:00:00            9   
5994          Closed    0 days 00:00:00      0 days 00:00:00           11   
5995          Closed    0 days 00:00:00      0 days 00:00:00           12   
5996          Closed    0 days 00:00:00      0 days 00:00:00            9   
5997          Closed    0 days 00:00:00      0 days 00:00:00           14   

      resolution_hour Resolution_class  total_response_seconds  
0            3.572778         Critical                   196.0  
1            0.553333             Good                   552.0  
2            1.538611             Fair                   588.0  
3            1.092500             Fair                   547.0  
4            0.555278             Good     

##### Formula to classify response time <= 10secs as good and > 10secs as bad

In [30]:
def convert_response_time(time_diff):
    if time_diff <= 10:
        return 'Good'
    elif time_diff > 10:
        return 'Bad'

creating a column based on the formula above:

In [31]:
data_service['response_time_status'] = data_service['total_response_seconds'].apply(convert_response_time)

In [32]:
data_service

Report ID Report Channel Customer Name State Key  \
0     AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1     AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2     AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3       AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4       AXA-20201231-1105-HSE           CH01          Arik    NGS003   
...                       ...            ...           ...       ...   
5993  AXA-20201204-7094-WiFi5           CH06        Lemway    NGS003   
5994  AXA-20201226-7095-WLESS           CH05        BetSan    NGS004   
5995  AXA-20201210-7096-WiFi6           CH06    ValueNaija    NGS029   
5996  AXA-20201208-7097-WLESS           CH03         Sahad    NGS007   
5997  AXA-20201201-7098-VOICE           CH01        BetSan    NGS018   

            Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0    2020-12-31 17:07:04.000 2020-12-31 17:10:20.000 2020-12-31 20:44:42.000   
1    2020-12-30 19:04:06.000 2020-12-30 19:13:18.000 2020-12-30 19:46:30.000   
2    2020-12-29 18:28:04.000 2020-12-29 18:37:52.000 2020-12-29 20:10:11.000   
3    2020-12-29 14:49:16.000 2020-12-29 14:58:23.000 2020-12-29 16:03:56.000   
4    2020-12-31 21:43:03.000 2020-12-31 21:47:39.000 2020-12-31 22:20:58.000   
...                      ...                     ...                     ...   
5993 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525   
5994 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377   
5995 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071   
5996 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601   
5997 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649   

         Fault Type  Operator           Ticket Close Time Ticket Status  \
0               NaN     Tunde                                    Active   
1     Customers End    Sekina         2021-01-01 13:27:17     Completed   
2     Customers End  Sherifat         2020-12-30 13:20:43     Completed   
3       Line damage  Kingsley         2020-12-29 21:34:31     Completed   
4     Customers End      Kola         2021-01-02 06:54:01     Completed   
...             ...       ...                         ...           ...   
5993    Line damage      Sheg  2020-12-04 12:53:29.525000     Completed   
5994  Customers End    Brenda  2020-12-26 16:58:54.377000     Completed   
5995  Customers End      Sola  2020-12-10 16:17:37.071000     Completed   
5996    Line damage  Jonathan  2020-12-08 13:39:54.601000     Completed   
5997  Customers End      Boye  2020-12-01 19:03:24.649000     Completed   

     Business Status Response_time_diff Resolution_time_diff  Hour_of_day  \
0               Open    0 days 00:03:16      0 days 03:34:22           17   
1             Closed    0 days 00:09:12      0 days 00:33:12           19   
2             Closed    0 days 00:09:48      0 days 01:32:19           18   
3             Closed    0 days 00:09:07      0 days 01:05:33           14   
4             Closed    0 days 00:04:36      0 days 00:33:19           21   
...              ...                ...                  ...          ...   
5993          Closed    0 days 00:00:00      0 days 00:00:00            9   
5994          Closed    0 days 00:00:00      0 days 00:00:00           11   
5995          Closed    0 days 00:00:00      0 days 00:00:00           12   
5996          Closed    0 days 00:00:00      0 days 00:00:00            9   
5997          Closed    0 days 00:00:00      0 days 00:00:00           14   

      resolution_hour Resolution_class  total_response_seconds  \
0            3.572778         Critical                   196.0   
1            0.553333             Good                   552.0   
2            1.538611             Fair                   588.0   
3            1.092500             Fair                   547.0   
4            0.555278             Good

In [33]:
data_service['response_time_status'].value_counts()

Bad     4945
Good    1053
Name: response_time_status, dtype: int64

#### Formula to easily filter tickets that were reported within 6-9pm and outside 6-9pm

In [34]:
def hour_of_day(row):
    if row >= 18 and row <= 21:
        return 'Within 6-9pm'
    else:
        return 'outside 6-9pm'

In [35]:
data_service['within_or_outside_6_9pm'] = data_service['Hour_of_day'].apply(hour_of_day)

In [36]:
data_service

Report ID Report Channel Customer Name State Key  \
0     AXA-20201231-1101-WLESS           CH01        Access    NGS001   
1     AXA-20201230-1102-WLESS           CH01        Access    NGS001   
2     AXA-20201229-1103-WLESS           CH01        Access    NGS001   
3       AXA-20201229-1104-HSE           CH01          Arik    NGS002   
4       AXA-20201231-1105-HSE           CH01          Arik    NGS003   
...                       ...            ...           ...       ...   
5993  AXA-20201204-7094-WiFi5           CH06        Lemway    NGS003   
5994  AXA-20201226-7095-WLESS           CH05        BetSan    NGS004   
5995  AXA-20201210-7096-WiFi6           CH06    ValueNaija    NGS029   
5996  AXA-20201208-7097-WLESS           CH03         Sahad    NGS007   
5997  AXA-20201201-7098-VOICE           CH01        BetSan    NGS018   

            Ticket Open Time        Ticket Resp Time          Issue Res Time  \
0    2020-12-31 17:07:04.000 2020-12-31 17:10:20.000 2020-12-31 20:44:42.000   
1    2020-12-30 19:04:06.000 2020-12-30 19:13:18.000 2020-12-30 19:46:30.000   
2    2020-12-29 18:28:04.000 2020-12-29 18:37:52.000 2020-12-29 20:10:11.000   
3    2020-12-29 14:49:16.000 2020-12-29 14:58:23.000 2020-12-29 16:03:56.000   
4    2020-12-31 21:43:03.000 2020-12-31 21:47:39.000 2020-12-31 22:20:58.000   
...                      ...                     ...                     ...   
5993 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525 2020-12-04 09:10:36.525   
5994 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377 2020-12-26 11:24:17.377   
5995 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071 2020-12-10 12:01:54.071   
5996 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601 2020-12-08 09:01:40.601   
5997 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649 2020-12-01 14:46:10.649   

         Fault Type  Operator           Ticket Close Time Ticket Status  \
0               NaN     Tunde                                    Active   
1     Customers End    Sekina         2021-01-01 13:27:17     Completed   
2     Customers End  Sherifat         2020-12-30 13:20:43     Completed   
3       Line damage  Kingsley         2020-12-29 21:34:31     Completed   
4     Customers End      Kola         2021-01-02 06:54:01     Completed   
...             ...       ...                         ...           ...   
5993    Line damage      Sheg  2020-12-04 12:53:29.525000     Completed   
5994  Customers End    Brenda  2020-12-26 16:58:54.377000     Completed   
5995  Customers End      Sola  2020-12-10 16:17:37.071000     Completed   
5996    Line damage  Jonathan  2020-12-08 13:39:54.601000     Completed   
5997  Customers End      Boye  2020-12-01 19:03:24.649000     Completed   

     Business Status Response_time_diff Resolution_time_diff  Hour_of_day  \
0               Open    0 days 00:03:16      0 days 03:34:22           17   
1             Closed    0 days 00:09:12      0 days 00:33:12           19   
2             Closed    0 days 00:09:48      0 days 01:32:19           18   
3             Closed    0 days 00:09:07      0 days 01:05:33           14   
4             Closed    0 days 00:04:36      0 days 00:33:19           21   
...              ...                ...                  ...          ...   
5993          Closed    0 days 00:00:00      0 days 00:00:00            9   
5994          Closed    0 days 00:00:00      0 days 00:00:00           11   
5995          Closed    0 days 00:00:00      0 days 00:00:00           12   
5996          Closed    0 days 00:00:00      0 days 00:00:00            9   
5997          Closed    0 days 00:00:00      0 days 00:00:00           14   

      resolution_hour Resolution_class  total_response_seconds  \
0            3.572778         Critical                   196.0   
1            0.553333             Good                   552.0   
2            1.538611             Fair                   588.0   
3            1.092500             Fair                   547.0   
4            0.555278             Good

##### save the cleaned data into a text/csv file;

In [37]:
#data_service.to_csv('updated_cleaned_dataset_nova.csv',index=False)